articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+[To discuss: Should there be a callback for when configuration is changed, e.g. altered parameters, new topics, etc.?]
+
+While in this state, the node will not receive any execution time to read topics, perform processing of data, respond to functional service requests, etc.
+
+In the inactive state, any data that arrives on topics will not be read. Whether or not data is queued will depend on the QoS parameters of the topics involved. [In other words, to queue data or not is up to each topic.]
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
```

Comment:
> Most errors should be handled at the system supervision level

I'm not sure I agree with this. Certainly it needs to be possible to handle errors at the supervisor level, but I wouldn't say that "most" errors will be handled there. In ROS 1 most all errors are resolved internal to the nodes. For example, if a device is not found on startup, a "nice" node would loop internally until the device is attached and report its state via a diagnostics message or via a topic which it doesn't publish to until its ready. There is only simple external error handling, like setting `respawn=true` in the launch file, so if the node shuts down, then `roslaunch` will relaunch it. This is typical for ROS 1, but I imagine that you're thinking more of a single executable in which a few "nodes" interact in a very tightly controller way. The author was probably thinking of the former when writing this.

There are two ways, in my opinion, to look at error handling at the "supervision" level. One would be in what we call the launch system, where you do things like look at node life cycles or maybe at return codes and react to those at a system wide level. This is the same kind of thing as `respawn=true` but more sophisticated. In this case nodes which share a process would likely be run by a generic piece of code like the "NodeletManager" in ROS 1.

The other way would be to write a program which loads the nodes, configures them externally, and then executes them, catching any errors directly and reacting by orchestrating them directly. Basically replacing the generic "NodeletManager" like thing with a custom manager which has domain specific knowledge. This is possible in ROS 1, but uncommon (I've never seen this), probably because it only would work with nodelets. Instead, system level error handling is done via nodes which monitor diagnostics and control other nodes via custom service calls or is done entirely though `roslaunch`.

With ROS 2, since execution is separated from the node, you can write a your own executor which has the opportunity to handle exceptions at the supervision level as you've described. So this will be easier to accomplish, but the question remains as to whether or not it should be done this way. Some system may need this to accomplish their goals, maybe with real-time or embedded systems. But my guess is that an improved version of what's possible in ROS 1 combined with a well defined life cycle interface is more scalable. The other thing to avoid when writing a custom orchestration process is breaking the encapsulation of a node by making it reliant on supervision level code in order to operate correctly.

I think it may be more complicated than this part of the document lets on.